crayfish/gui/mesh_calculator_dialog.py

```diff
@@ -149,15 +156,56 @@ def add_dataset_to_layer(self):
     def on_dataset_output_filename_changed(self, dummy):
         self.set_accept_button_state()
 
+    def list_polygon_layers(self):
+        return [layer for layer in QgsMapLayerRegistry.instance().mapLayers().values() if
+                layer.type() == QgsMapLayer.VectorLayer and layer.geometryType() == QGis.Polygon]
+
+    def combine_geometries(self, features):
+        geoms = None
+        for feat in features:
+            geom = feat.geometry()
+            if not geom: continue
+
+            err = geom.validateGeometry()
+            if not err:
+                if not geoms:
+                    geoms = geom
+                else:
+                    geoms = geoms.combine(geom)
+
+        return geoms
+
+    def toogle_extend_mask(self):
+        if self.useMaskCb.checkState() == Qt.Checked:
+            self.extendBox.setVisible(False)
+            self.maskBox.setVisible(True)
+        else:
+            self.extendBox.setVisible(True)
+            self.maskBox.setVisible(False)
+
     def on_accept_clicked(self):
         mesh = self.layer.mesh
-        success = mesh.create_derived_dataset(
-            expression=self.formula_string(),
-            time_filter=self.time_filter(),
-            spatial_filter=self.spatial_filter(),
-            add_to_mesh=self.add_dataset_to_layer(),
-            output_filename=self.output_filename()
-        )
+
+        success = None
+        if self.useMaskCb.checkState() == Qt.Checked:
+            mask_layer = self.cboLayerMask.currentLayer()
+            feats = list(mask_layer.getFeatures())
+            if mask_layer and feats:
```

Comment:
no else statement

also, can you create function like "create_mask_geoms_wkt", so the logic is simpler? It is kind of hard to read this function